### PR TITLE
Only compute potentials at beginTimeStep after an event

### PR DESCRIPTION
**Pull request by totto82**
This will need more testing, but is looks promising.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
Test time this PR:
113/129 Test #102: compareECLFiles_flow+MOD4_UDQ_ACTIONX .............................***Failed   55.65 sec
118/129 Test #101: compareECLFiles_flow+MOD4_GRP .....................................***Failed   61.43 sec
Total Test time (real) = 363.46 sec

Test time from https://github.com/OPM/opm-simulators/pull/2975 (master)
118/129 Test #102: compareECLFiles_flow+MOD4_UDQ_ACTIONX .............................   Passed  108.42 sec
119/129 Test #101: compareECLFiles_flow+MOD4_GRP .....................................   Passed  139.33 sec
Total Test time (real) = 636.67 sec

There are some real failures among the tests cases that I need to fix, but the speed up real. I see speed up on bigger models as well. But the relative difference is of course much smaller since the wells does not dominate the runtime on real models.

I also think this PR fix a real bug. In the current code the try-catch is outside the loop. Doesn't the loop then terminate on the first failure?

**Comment by blattms** [comment]
> I also think this PR fix a real bug. In the current code the try-catch is outside the loop. Doesn't the loop then terminate on the first failure?

Indeed, all remaining wells would be skipped.

**Comment by totto82** [comment]
> Indeed, all remaining wells would be skipped.
I have a small hope that this could explain why all MSW cases fails locally. I will run a pre update_results and check it locally

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by totto82** [comment]
> I have a small hope that this could explain why all MSW cases fails locally. I will run a pre update_results and check it locally

It didn't do the trick. I will cherry-pick the  bug fix and make a new PR to make it easier to review.

**Comment by totto82** [comment]
This will need to changes in opm-common OPM/opm-common#2167

**Comment by totto82** [comment]
jenkins build this opm-common=2167 please

**Comment by totto82** [comment]
jenkins build this opm-common=2167 please

**Comment by totto82** [comment]
jenkins build this opm-common=2167 please

**Comment by totto82** [comment]
jenkins build this opm-common=2167 please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                 | Relative |
| ------- | --------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1    | 1.007    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8    | 1.012    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1    | 1.003    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8    | 1.003    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c - Threads: 1 | 0.888    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c - Threads: 8 | 1.033    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1052

**Comment by totto82** [comment]
jenkins build this please

**Comment by blattms** [comment]
Sorry for hijacking:
What I have been wondering is how fancy or Event system is. I am aware that if  a statement changing the well specification (e.g. WCONINJE or similar) happens there will be an event. But recently there was a case where one of the rates specified in WCONINJE was a function depending on some other (possibly changing) properties. I am pretty sure that when querying that value we will get the correct one, but I am not sure whether it will trigger an event.

Maybe having a developer option for always querying the wells from ECL regardless from events might help in debugging

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 1.037    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1.014    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 1.045    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.004    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 1.28     |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 0.784    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1130

**Comment by totto82** [comment]
jenkins build this please

File: opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp [hunk 1]
```diff
@@ -264,12 +264,6 @@ namespace Opm
                             wellReservoirRates()[ idx ] = prevState->wellReservoirRates()[ oldidx ];
                         }
 
-                        // Well potentials
-                        for( int i=0, idx=newIndex*np, oldidx=oldIndex*np; i<np; ++i, ++idx, ++oldidx )
-                        {
-                            wellPotentials()[ idx ] = prevState->wellPotentials()[ oldidx ];
-                        }
-
```
Comment (bska) [inline opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp:272]: Is this a mismerge or did you intend to remove this block of code?

Comment (totto82) [reply opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp:272]: Yes it is  mismerge. I will update asap.

**Comment by totto82** [comment]
jenkins build this please

**Comment by bska** [comment]
I haven't looked closely at this, nor reviewed the failing regression tests.  I'll just mention that in the current PR, the [`nw`](https://github.com/OPM/opm-simulators/blob/f4ef30eb8a930ad3dec2809f74e598ffdf6f0e97/opm/simulators/wells/BlackoilWellModel_impl.hpp#L1531) object in [`BlackoilWellModel::updateWellPotentials()`](https://github.com/OPM/opm-simulators/blob/f4ef30eb8a930ad3dec2809f74e598ffdf6f0e97/opm/simulators/wells/BlackoilWellModel_impl.hpp#L1527-L1528) becomes unused.  That generates a number of [warnings](https://ci.opm-project.org/job/opm-simulators-PR-builder/2136/gcc/fileName.1685820536/) which should be easy to remove.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 0.978    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1.005    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 0.97     |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.004    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 0.919    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 0.786    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1175

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by bska** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 1.003    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1.002    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 1.004    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 0.97     |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 0.999    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 0.825    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1196

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
The usual noise with he MSW makes a lot a test failures. I have done a manual check and they are all minor. I suggest we do a update of the reference data and get the code in. Assuming the code itself is ok. 

    mpi.compareECLFiles_flow+2D_H__ ok
    mpi.compareECLFiles_flow+2D_POLYMER_INJECTIVITY ok
    mpi.compareECLFiles_flow+BASE_MODEL_1 ok
    mpi.compareECLFiles_flow+MSW_MODEL_1 ok
    mpi.compareECLFiles_flow+0A1_GRCTRL_LRAT_ORAT_BASE_MODEL2_MSW ok
    mpi.compareECLFiles_flow+0A2_GRCTRL_LRAT_ORAT_GGR_BASE_MODEL2_MSW ok
    mpi.compareECLFiles_flow+0A3_GRCTRL_LRAT_LRAT_BASE_MODEL2_MSW ok
    mpi.compareECLFiles_flow+0A4_GRCTRL_LRAT_LRAT_GGR_BASE_MODEL2_MSW ok 
    mpi.compareECLFiles_flow+0_BASE_MODEL2 (66 != 67)
    mpi.compareECLFiles_flow+ACTIONX_M1 ok
    mpi.compareECLFiles_flow+1_MULTREGT_MODEL2 ok 
    mpi.compareECLFiles_flow+2_MULTXYZ_MODEL2 (61 != 57)
    mpi.compareECLFiles_flow+4_MINPVV_MODEL2 (62 != 61)
    mpi.compareECLFiles_flow+3_MULTFLT_MODEL2 (62 != 61)
    mpi.compareECLFiles_flow+5_SWATINIT_MODEL2 ok
    mpi.compareECLFiles_flow+6_ENDSCALE_MODEL2 (60 != 62)
    mpi.compareECLFiles_flow+7_HYSTERESIS_MODEL2 (64 != 62)
    mpi.compareECLFiles_flow+8_MULTIPLY_TRANXYZ_MODEL2
    mpi.compareECLFiles_flow+9_1A_DEPL_MAX_RATE_MIN_BHP_MSW ok
    mpi.compareECLFiles_flow+9_2A_DEPL_GCONPROD_1L_MSW (33 != 31)
    mpi.compareECLFiles_flow+9_2B_DEPL_GCONPROD_2L_MSW ok
    mpi.compareECLFiles_flow+9_1B_DEPL_MAX_RATE_MIN_THP_MSW ok
    mpi.compareECLFiles_flow+9_3A_GINJ_REIN-G_MSW  (31 != 30)
    mpi.compareECLFiles_flow+9_3B_GINJ_GAS_EXPORT_MSW (31 != 30)
    mpi.compareECLFiles_flow+9_3C_GINJ_GAS_GCONSUMP_MSW ok
    mpi.compareECLFiles_flow+9_3E_GAS_MIN_EXPORT_MSW  ok
    mpi.compareECLFiles_flow+9_3D_GINJ_GAS_MAX_EXPORT_MSW ok
    mpi.compareECLFiles_flow+9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_MSW ok
    mpi.compareECLFiles_flow+9_4B_WINJ_VREP-W_MSW ok
    mpi.compareECLFiles_flow+9_4C_WINJ_GINJ_VREP-W_REIN-G_MSW ok
    mpi.compareECLFiles_flow+1_MSW_MODEL6 ok
    mpi.compareECLFiles_flow+9_4D_WINJ_GINJ_GAS_EXPORT_MSW ok
    mpi.compareECLFiles_flow+3D_TRAN_OPERATOR ok
    mpi.compareECLFiles_flow+MOD4_GRP (105 != 94)
    mpi.compareECLFiles_flow+MOD4_UDQ_ACTIONX (94 != 98)

**Comment by totto82** [comment]
@bska Any comment on the code? If you approve the code changes I want to start the updating of the reference solution process.

**Review by bska — APPROVED** [review]
The code is okay in isolation.  The repeated keyword lookup in `SummaryConfig` could possibly be factored out to reduce the cost, but that's a subject of a possible follow-up PR.  Other than that, I think we can update the reference solutions and merge this.

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=535 please

**Comment by bska** [comment]
The updated reference solutions are installed on the CI system.  I'll merge this into master now.

**Pull request merged**